### DISCUSSION TITLE
Still need cache-busting on CSS file?

### DISCUSSION THREAD

**Discussion by szalapski** [dotnet/aspnetcore#27657] [Q&A] [answered]
I [read this](https://github.com/dotnet/aspnetcore/issues/17036
) but still am having trouble:  
I have my Blazor WebAssembly site, with a index.html file that bootstraps Blazor in a normal way.   In its `<head>`, I have `  <link href="css/app.min.css" rel="stylesheet" />`.  That file is built from SCSS using a Dockerfile RUN step on deployment.

According to [that issue](https://github.com/dotnet/aspnetcore/issues/17036) and @SteveSandersonMS , I don't need cache-busting URLs, as Blazor will automatically request a fresh one as needed.  Yet when I deploy my site and open it in Firefox, I get the old styles and see in Firefox developer tools network tab that app.min.css is retrieved from disk cache.  I have to press Ctrl+F5 to get the new styles.

So what is going wrong in this chain?  It seems to me that I still need cache-busting URL for my CSS on each deployment.

**Comment by szalapski** [comment]
Cross-posted here: https://stackoverflow.com/questions/64760377/still-need-cache-busting-on-css-file-in-blazor-webassembly-site

**Comment by SteveSandersonMS** [comment, marked as answer]
Whether or not the browser uses cached copies is controlled by the HTTP headers sent by your server when it returned the `.css` file previously. As long as it has a valid `etag` then the browser should check for updates via a request with `an If-Modified-Since` header. This isn't something that Blazor WebAssembly controls, since Blazor WebAssembly is purely client-side and hence doesn't control the headers sent by your server.

**Reply by szalapski** [reply]
Am I correct that an etag-based solution would still require a app.min.css request on every load of index.html?  It would just often return a 304 without content, rather than a 200, right?    I think I want to avoid that request--is this a reasonable desire?  Isn't the point of browser caching to avoid making requests at all?  Not arguing, I just want to make sure I'm not missing something here.

**Reply by szalapski** [reply]
To reiterate, if there is a `<link>` to app.min.css in index.html (the static page that serves up Blazor in the first place), what you say here wouldn't be true anymore, right? That is, the browser will retrieve app.min.css regardless of anything Blazor.  This will result in unnecessary requests by default, or possibly-stale cache hits if CacheControl is enabled.  Am I supposed to take my CSS references out of index.html, and if so, where do I reference app.min.css?

**Reply by szalapski** [reply]
I tried it as you suggest, removing the CacheControl, and I do indeed see that _sometimes_ the_ .CSS is not requested after the first load, but other times it is (with an etag resulting in a 304).

The first scenario is may main question, where it does not request app.min.css at all.

![image](https://user-images.githubusercontent.com/2723505/99535978-5212bb00-296f-11eb-9025-b910923c3a69.png)

How can this be, since there is a `<link>` to it in my index.html that is unrelated to Blazor?  Does Blazor somehow take over and cause the page not to request the `<link>` in its `<head>`?

For the second scenario, I know strictly speaking we cannot guarantee that the browser won't request the asset...is this just Firefox applying some kind of heuristic, like perhaps if Ihe user waits awhile between refreshes, it "tries a little harder" to be helpful and thus runs an extraneous request with an etag?  Just speculating, but I definitely see a request with a 304 response on this CSS file sporadically.

**Reply by SteveSandersonMS** [reply]
Nope, Blazor doesn’t control this. These are native browser/HTTP behaviours.

**Reply by szalapski** [reply]
Okay, so the most ordinary way of referencing a CSS file in a static html file, if CacheControl is on (and we want to avoid these etag-304 requests), would still need cache-busting--but there is no reasonable way to do this since index.html is static, right?  

How do you reference your main site CSS file **with CacheControl on** in your Blazor sites?  Or do you always just incur the penalty of letting the browser making the etagged requests?   Begging your patience, and thank you for taking the time.

**Reply by SteveSandersonMS** [reply]
The etag requests aren't typically significant considering HTTP pipelining.

However these are really general questions about browser behavior and server configuration, so for more details I'd recommend going to external docs and guidance ([example](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control)). Hope that's OK!

**Reply by szalapski** [reply]
OK, so I went back to the simple approach you suggest: simply `app.UseStaticFiles()` without any Cache-Control settings.  As I'm sure you know, this gives me "no-cache" but not "no-content", and I expect **always** to see an etag request that results in a 204 when unchanged or a 200 when changed.  **Instead**, what I see is that _sometimes_ there is no CSS request at all, and the browser just uses its cached version without checking.  Other times, there is a CSS request.  (I am testing with Firefox.)

Thus, when I deploy a new version of the CSS, users get display bugs shown on their page.

This seems to be in contradiction to the MDN page you cite above, which says:

> However, the stored response MUST always go through validation with the origin server first before using it.

Yet that doesn't always happen.  I'll post some screenshots...

I also see [this comment](https://github.com/dotnet/aspnetcore/issues/16344#issuecomment-379208671) from you... curious, has your position changed?  Because I believe I'm looking for exactly what you suggest:

> Longer term, it would be good if our publishing process generated content-hashed URLs. Then we wouldn't need the browser to check for Etag-based modifications at all. It could just cache all compiled-file responses forever and not check with the server at all.

**Reply by szalapski** [reply]
Here are two consecutive requests, once pressing F5 gave me an unexpected no-request-at-all reresh, and then another simple F5 refresh then gave me the expected CSS request with the expected 304.  In the first case, the user could get an improperly-styled page.  Are you sure there's nothing in Blazor that could be suppressing certain requests?  When I hit F5, does Blazor somehow avoid making a completely new request, or something like that?

![image](https://user-images.githubusercontent.com/2723505/101778785-efb66180-3ab9-11eb-8a4c-a80f780c261a.png)
![image](https://user-images.githubusercontent.com/2723505/101778711-d6adb080-3ab9-11eb-9728-63a8792eece4.png)

**Reply by SteveSandersonMS** [reply]
My comment was about the framework resources that Blazor itself controls the responses for. For those, the framework chooses and knows about the desired refresh semantics. Since I posted that, we've implemented logic within `blazor.*.js` that uses the browser's `Cache` APIs to keep track of content-hash-keyed data and hence avoid the need to check for updates, so in effect it's already done.

For other resources that Blazor doesn't control, such as your CSS files, it's entirely up to you to choose the desired caching semantics and to control that via HTTP headers or any custom logic you want around loading them. In the vast majority of cases developers don't need to take any steps, because ASP.NET Core's built-in default static files middleware produces `Etag` headers that provide a good balance of using the browser's cache well. But of course you can override it if you want as per [static file middleware docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-5.0).

**Reply by szalapski** [reply]
OK, thanks for all the explanation.  I agree--that all makes sense and should be correct.  The Firefox behavior seems odd to me, so [I am asking about that here](https://stackoverflow.com/questions/65237363/what-is-the-exact-behavior-of-firefox-on-a-page-refresh-regarding-a-cached-asset).

**Reply by szalapski** [reply]
I now suspect my problem might actually be [a bug or design issue in Firefox; we'll see if we can pursue it further there](https://support.mozilla.org/en-US/questions/1317462).

**Comment by SteveSandersonMS** [comment]
> Are you sure there's nothing in Blazor that could be suppressing?

No, Blazor is not involved in serving the static content (CSS files in `wwwroot`) at all.
